### PR TITLE
Message Queue Architecture and Options

### PR DESCRIPTION
**Pull request by tariqjawed83**
## Problem

Currently Magento uses MySQL and RabbitMQ for Message Queue implementations, many new technologies has came-along which can scale and managed much better in the cloud based environment.

## Solution

This proposal is looking into current messaging architecture and potential candidates as additional technology adapter for messaging platform in addition to MySQL and RabbitMQ; specially looking into managed cloud based queueing technologies. 

## Requested Reviewers

@nuzil @paliarush @maghamed  @buskamuza @hostep @naydav

### CODE REVIEW

**Comment by kandy** [comment]
Can you add Azure and GCP Message Queue Services to overview?
Also will be nice if you review integration with Kubernetes

**Comment by tariqjawed83** [comment]
@kandy thanks for your comments, I try to add some Azure technology comparison as well. There is separate stream of work for Kubernetes going on though.

File: design-documents/message-queue/message-queue-architecture.md [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+| Implementation | Description                                                  |
+| -------------- | ------------------------------------------------------------ |
+| MySQL          | This is based on MySQL database, currently it doesn't support the Bulk APIs; this doesn't scale well and designed for out of the box experience; for practical purposes RabbitMQ should be utilized to handle large scale requirements |
+| RabbitMQ       | RabbitMQ implementations is based on AMQP protocol, which supports callback mechanism and polling based approach as its supported by underlying channel implementation & specification |
+
+![Queue Interface UML Diagram](QueueInterfaceUML.png)
+
+### Consumers Runner Process
+
+ConsumersRunner is the main class, which facilitates the processing of Queued Messages, and based on the configuration in **queue_consumer.xml**, it calls the appropriate ConsumerInterface implementation.
```
[the remark this answers is not in the document]
Comment (tariqjawed83) [reply design-documents/message-queue/message-queue-architecture.md]: sure will do that,

File: design-documents/message-queue/message-queue-architecture.md [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+
+This QueueInterface is been utilized by Consumer classes, there are currently three different implementations of ConsumerInterface
+
+|      | Class Name    | Purpose / Description                                        |
+| ---- | ------------- | ------------------------------------------------------------ |
+| 1    | Consumer      | This class is for both synchronous and asynchronous style message processing |
+| 2    | MassConsumer  | This is mainly designed for asynchronous style message processing; primarily used for Async APIs and Async Bulk APIs |
+| 3    | BatchConsumer | This class supports batch processing of messages, helps in picking & merging the messages to a specified batch size, and then process them together, before querying another batch |
+
+![Queue Interface UML Diagram](ConsumerInterfaceUML.png)
```
[the remark this answers is not in the document]
Comment (tariqjawed83) [reply design-documents/message-queue/message-queue-architecture.md]: MassConsumer is handling Async Callbacks only and the logic is decomposed in MassConsumer better; whereas Consumer is handling both cases sync/async based on topic type; sync topics get a response from callbacks in reply_to queue

File: design-documents/message-queue/AWSEventBridge.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
+#### 1- AWS EventBridge
```
Comment (nuzil) [inline design-documents/message-queue/AWSEventBridge.md]: Btw there are already some project, maybe you can reference on it
https://github.com/magento-engcom/amazon-event-bridge

File: design-documents/message-queue/AWSMQ.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+Since its a managed service, it provides multi zone fault tolreance and resiliancy out of the box.
+
+<img src="AWSMQArchitectureNew.png" alt="image-20190927212814114" width="50%" />
+
+
+
+##### Evaluation Table - Details
+
+Most of the features are available since Magento is also using AMQP protocol with RabbitMQ, but the protocol version is different, RabbitMQ uses 0.9 and Amazon MQ is using AMQP 1.0; so any migration would require porting of Queues from RabbitMQ to AWS MQ. There might be some changes and adjustments to the QueueInterface implementation code to accomodate the new protocol differences. AMQP 1.0 is a completely different protocol.
```
Comment (nuzil) [inline design-documents/message-queue/AWSMQ.md]: Not sure its easiest task, related to backward compatibility. Maybe even make sense to make its as separate module. This part have to be checked

**Comment by nuzil** [comment]
Looks great!

File: design-documents/message-queue/PHPEnqueueLibrary.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+| queue_consumer.xml         | It defines the consumer processes and their attributes (like queue name, connection, maxMessages, consumerInstance type and callback handler etc.)<br />ConsumerInstance type defines what kind of Consumer you want to use in Magento as defined in [Different Types of Consumers](MagentoMessageQueueDesign.md#Different-Types-of-Consumers) | N/A                            | N/A                                                          |
+| queue_topology.xml         | Defines the exchange where Topic is been published (Topic to Queue mapping) | N/A                            | N/A                                                          |
+| queue_publisher.xml        | It defines the publishers and its attributes (connection name and exchange) | N/A                            | N/A                                                          |
+
+
+
+### Open Questions?
+
+1. There are two places where callback handlers are defined for a Topic/Queue. One in communications.xml; and other in queue_consumer.xml, which one has priority over other?
```
Comment (nuzil) [inline design-documents/message-queue/PHPEnqueueLibrary.md]: So about this one
Answer on this question depends from Consumer implementation
1) `Magento\Framework\MessageQueue\Consumer` and `Magento\Framework\MessageQueue\BatchConsumer` - this one using Handlers from "queue_consumer.xml" is set. If there are no one set, then from "communications.xml"
2) `Magento\AsynchronousOperations\Model\MassConsumer` - This one use both. First from queue_consumer and then result will be transferred to handler from communications.xml

File: design-documents/message-queue/PHPEnqueueLibrary.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+<img src="legend_img.png" alt="Legend" width="70%" height="70%" />
+
+
+
+## SimpleClient & Queue Interoperablity
+
+There is a Magento2 module available for Enqueue library, [Magento 2 Enqueue Quick Tour](https://github.com/php-enqueue/enqueue-dev/blob/master/docs/magento2/quick_tour.md), this extension provides easy to configure Magento Admin screen, which setup any supported Queue technology within Enqueue (behind the scenes it add configuration here [Magento 2 Enqueue Config Example](https://github.com/php-enqueue/magento2-enqueue/blob/master/etc/config.xml)). In the end, it makes it straightforward to publish and consume messages in your PHP application.
+
+Magento already implements lot of functionality by itself (for instance keeping track of Connection, Queues and their callbacks, consumer types etc.). We can potentially use only the subset of Enqueue library to interface with the undelying Messaging System in a uniform manner; by providing another QueueInterface implemention for **Enqueue** module.
```
Comment (nuzil) [inline design-documents/message-queue/PHPEnqueueLibrary.md]: I guess its a correct approach. Implementing such library will require a lot of time and deep integration in existed implementation. I would not use existed extension and would go to implementation of the subset of Enqueue library

File: design-documents/message-queue/PHPEnqueueLibrary.md (hunk 5 above)
Comment (tariqjawed83) [reply design-documents/message-queue/PHPEnqueueLibrary.md]: thanks for the feedback, this is helpful

**Pull request merged**